File: src/main/detect.ts
// Infer run/test/build (and docs/CLAUDE.md paths) from an existing project's structure.
// Best-effort and conservative: only emit a value when a recognised marker is found.
// The renderer presents these as editable defaults — never silently authoritative.

import { readFileSync, statSync } from 'node:fs'
import { join } from 'node:path'
import type { DetectedSettings } from '../shared/types'

function isDir(p: string): boolean {
  try {
    return statSync(p).isDirectory()
  } catch {
    return false
  }
}

function isFile(p: string): boolean {
  try {
    return statSync(p).isFile()
  } catch {
    return false
  }
}

/** Pick the JS package manager from whichever lockfile is present. */
function jsManager(root: string): string {
  if (isFile(join(root, 'pnpm-lock.yaml'))) return 'pnpm'
  if (isFile(join(root, 'yarn.lock'))) return 'yarn'
  if (isFile(join(root, 'bun.lockb'))) return 'bun'
  return 'npm'
}

function runScript(mgr: string, script: string): string {
  // `npm run X` / `pnpm run X` / `yarn X` / `bun run X`
  return mgr === 'yarn' ? `yarn ${script}` : `${mgr} run ${script}`
}

function fromPackageJson(root: string, out: DetectedSettings): boolean {
  const pkgPath = join(root, 'package.json')
  if (!isFile(pkgPath)) return false
  let scripts: Record<string, string> = {}
  try {
    const raw = readFileSync(pkgPath, 'utf8')
    const pkg = JSON.parse(raw) as { scripts?: Record<string, string> }
    scripts = pkg.scripts ?? {}
  } catch {
    return false
  }
  const mgr = jsManager(root)
  out.sources.push(`package.json (${mgr})`)

  const runKey = ['dev', 'start', 'serve'].find((k) => scripts[k])
  if (runKey) out.runCommand = runScript(mgr, runKey)

  if (scripts.test) out.testCommand = mgr === 'yarn' ? 'yarn test' : `${mgr} test`

  const buildKey = ['build', 'compile'].find((k) => scripts[k])
  if (buildKey) out.buildCommand = runScript(mgr, buildKey)
  return true
}

function fromMakefile(root: string, out: DetectedSettings): boolean {
  const mk = join(root, 'Makefile')
  if (!isFile(mk)) return false
  const targets = new Set<string>()
  try {
    const text = readFileSync(mk, 'utf8')
    for (const line of text.split('\n')) {
      const m = /^([A-Za-z0-9_.-]+)\s*:/.exec(line)
      if (m) targets.add(m[1])
    }
  } catch {
    return false
  }
  out.sources.push('Makefile')
  if (!out.runCommand && targets.has('run')) out.runCommand = 'make run'
  if (!out.testCommand && targets.has('test')) out.testCommand = 'make test'
  if (!out.buildCommand && targets.has('build')) out.buildCommand = 'make build'
  return true
}

function fromCargo(root: string, out: DetectedSettings): boolean {
  if (!isFile(join(root, 'Cargo.toml'))) return false
  out.sources.push('Cargo.toml')
  out.runCommand ??= 'cargo run'
  out.testCommand ??= 'cargo test'
  out.buildCommand ??= 'cargo build'
  return true
}

function fromPython(root: string, out: DetectedSettings): boolean {
  const py = isFile(join(root, 'pyproject.toml'))
    ? 'pyproject.toml'
    : isFile(join(root, 'setup.py'))
      ? 'setup.py'
      : ''
  if (!py) return false
  out.sources.push(py)
  out.testCommand ??= 'pytest'
  return true
}

function fromGo(root: string, out: DetectedSettings): boolean {
  if (!isFile(join(root, 'go.mod'))) return false
  out.sources.push('go.mod')
  out.runCommand ??= 'go run .'
  out.testCommand ??= 'go test ./...'
  out.buildCommand ??= 'go build ./...'
  return true
}

const COMPOSE_FILES = ['docker-compose.yml', 'docker-compose.yaml', 'compose.yml', 'compose.yaml']

function fromCompose(root: string, out: DetectedSettings): boolean {
  if (!COMPOSE_FILES.some((f) => isFile(join(root, f)))) return false
  out.sources.push('docker compose')
  // Fill run/build only if no richer toolchain already did — compose is the fallback.
  out.runCommand ??= 'docker compose up'
  out.buildCommand ??= 'docker compose build'
  return true
}

export function detectProjectSettings(root: string): DetectedSettings {
  const out: DetectedSettings = { sources: [] }
  if (!root || !isDir(root)) return out

  // JS toolchain wins for run/test/build; other stacks fill gaps they recognise.
  fromPackageJson(root, out)
  fromMakefile(root, out)
  fromCargo(root, out)
  fromGo(root, out)
  fromPython(root, out)
  fromCompose(root, out) // fallback run/build for compose-only projects

  // Surface .env so the user can see/edit the variables (they're also auto-merged at spawn).
  const envPath = join(root, '.env')
  if (isFile(envPath)) {
    try {
      if (statSync(envPath).size <= 256 * 1024) {
        const text = readFileSync(envPath, 'utf8').trim()
        if (text) {
          out.env = text
          out.sources.push('.env')
        }
      }
    } catch {
      /* unreadable .env — ignore */
    }
  }

  if (isFile(join(root, 'CLAUDE.md'))) {
    out.claudeMdPath = './CLAUDE.md'
    out.sources.push('CLAUDE.md')
  }
  if (isDir(join(root, 'docs'))) {
    out.docsPath = 'docs/'
    out.sources.push('docs/')
  }

  return out
}

File: src-tauri/src/detect.rs
// Infer run/test/build (+ docs/CLAUDE.md/.env) from an existing project's structure.
// Mirror of src/main/detect.ts — conservative, only emits recognised markers.

use std::path::Path;

use crate::types::DetectedSettings;

fn is_file(root: &str, name: &str) -> bool {
    Path::new(root).join(name).is_file()
}
fn is_dir(root: &str, name: &str) -> bool {
    Path::new(root).join(name).is_dir()
}

fn js_manager(root: &str) -> &'static str {
    if is_file(root, "pnpm-lock.yaml") {
        "pnpm"
    } else if is_file(root, "yarn.lock") {
        "yarn"
    } else if is_file(root, "bun.lockb") {
        "bun"
    } else {
        "npm"
    }
}

fn run_script(mgr: &str, script: &str) -> String {
    if mgr == "yarn" {
        format!("yarn {script}")
    } else {
        format!("{mgr} run {script}")
    }
}

pub fn detect(root: &str) -> DetectedSettings {
    let mut out = DetectedSettings::default();
    if root.is_empty() || !Path::new(root).is_dir() {
        return out;
    }

    // package.json
    if is_file(root, "package.json") {
        if let Ok(text) = std::fs::read_to_string(Path::new(root).join("package.json")) {
            if let Ok(pkg) = serde_json::from_str::<serde_json::Value>(&text) {
                let mgr = js_manager(root);
                out.sources.push(format!("package.json ({mgr})"));
                let scripts = pkg.get("scripts").and_then(|s| s.as_object());
                let has = |k: &str| scripts.map(|s| s.contains_key(k)).unwrap_or(false);
                for k in ["dev", "start", "serve"] {
                    if has(k) {
                        out.run_command = Some(run_script(mgr, k));
                        break;
                    }
                }
                if has("test") {
                    out.test_command = Some(if mgr == "yarn" {
                        "yarn test".into()
                    } else {
                        format!("{mgr} test")
                    });
                }
                for k in ["build", "compile"] {
                    if has(k) {
                        out.build_command = Some(run_script(mgr, k));
                        break;
                    }
                }
            }
        }
    }

    // Makefile
    if is_file(root, "Makefile") {
        if let Ok(text) = std::fs::read_to_string(Path::new(root).join("Makefile")) {
            out.sources.push("Makefile".into());
            let targets: Vec<String> = text
                .lines()
                .filter_map(|l| {
                    l.split_once(':').and_then(|(t, _)| {
                        let t = t.trim();
                        if !t.is_empty()
                            && t.chars().all(|c| c.is_alphanumeric() || "._-".contains(c))
                        {
                            Some(t.to_string())
                        } else {
                            None
                        }
                    })
                })
                .collect();
            if out.run_command.is_none() && targets.iter().any(|t| t == "run") {
                out.run_command = Some("make run".into());
            }
            if out.test_command.is_none() && targets.iter().any(|t| t == "test") {
                out.test_command = Some("make test".into());
            }
            if out.build_command.is_none() && targets.iter().any(|t| t == "build") {
                out.build_command = Some("make build".into());
            }
        }
    }

    if is_file(root, "Cargo.toml") {
        out.sources.push("Cargo.toml".into());
        out.run_command.get_or_insert("cargo run".into());
        out.test_command.get_or_insert("cargo test".into());
        out.build_command.get_or_insert("cargo build".into());
    }
    if is_file(root, "go.mod") {
        out.sources.push("go.mod".into());
        out.run_command.get_or_insert("go run .".into());
        out.test_command.get_or_insert("go test ./...".into());
        out.build_command.get_or_insert("go build ./...".into());
    }
    if is_file(root, "pyproject.toml") || is_file(root, "setup.py") {
        out.sources.push(
            if is_file(root, "pyproject.toml") {
                "pyproject.toml"
            } else {
                "setup.py"
            }
            .into(),
        );
        out.test_command.get_or_insert("pytest".into());
    }
    if COMPOSE_PRESENT.iter().any(|f| is_file(root, f)) {
        out.sources.push("docker compose".into());
        out.run_command.get_or_insert("docker compose up".into());
        out.build_command
            .get_or_insert("docker compose build".into());
    }

    // .env surfaced
    if is_file(root, ".env") {
        if let Ok(meta) = std::fs::metadata(Path::new(root).join(".env")) {
            if meta.len() <= 256 * 1024 {
                if let Ok(text) = std::fs::read_to_string(Path::new(root).join(".env")) {
                    let t = text.trim();
                    if !t.is_empty() {
                        out.env = Some(t.to_string());
                        out.sources.push(".env".into());
                    }
                }
            }
        }
    }

    if is_file(root, "CLAUDE.md") {
        out.claude_md_path = Some("./CLAUDE.md".into());
        out.sources.push("CLAUDE.md".into());
    }
    if is_dir(root, "docs") {
        out.docs_path = Some("docs/".into());
        out.sources.push("docs/".into());
    }
    out
}

const COMPOSE_PRESENT: &[&str] = &[
    "docker-compose.yml",
    "docker-compose.yaml",
    "compose.yml",
    "compose.yaml",
];
